corehq/apps/app_manager/tasks.py

```diff
@@ -91,3 +93,54 @@ def update_linked_app_and_notify_task(domain, app_id, master_app_id, user_id, em
 def load_appcues_template_app(domain, username, app_slug):
     from corehq.apps.app_manager.views.apps import load_app_from_slug
     load_app_from_slug(domain, username, app_slug)
+
+
+@task(queue='background_queue', ignore_result=True)
+def analyse_new_app_build(domain, new_build_id):
+    new_build = get_app(domain, new_build_id)
+
+    check_for_custom_callouts(new_build)
+    check_build_dependencies(new_build)
+
+
+def check_for_custom_callouts(new_build):
+    from corehq.apps.app_manager.util import app_callout_templates_ids
+    template_ids = app_callout_templates_ids()
+
+    def app_has_custom_intents():
+        return any(
+            any(set(form.wrapped_xform().odk_intents) - template_ids)
+            for form in new_build.get_forms()
+        )
+
+    if app_has_custom_intents():
+        metrics_counter(
+            'commcare.app_build.custom_app_callout',
+            tags={'domain': new_build.domain, 'app_id': new_build.copy_of},
+        )
+
+
+def check_build_dependencies(new_build):
+    """
+    Reports whether the app dependencies have been added or removed.
+    """
+
+    def has_dependencies(build):
+        return bool(
+            build.profile.get('features', {}).get('dependencies')
+        )
+
+    new_build_has_dependencies = has_dependencies(new_build)
+    app_build_ids = get_build_ids(new_build.domain, new_build.copy_of)
```

Comment:
I'd recommend checking [this](https://github.com/dimagi/commcare-hq/blob/b24fc4411c6c9c90446740f4f2731891d9849380/corehq/apps/app_manager/views/releases.py#L134) out and using this couch view to fetch directly the relevant version and then one before it.
Though good to test and see how it will be used here.

Reply:
Will do!

Reply:
@mkangia I've tested the approach you suggested, but since it uses a couch view the `modules` is filtered out and not returned from the raw app doc, so I can't access the forms which I need for this processing. I didn't find any other view which currently returns the modules, so I think at this point I'm going to leave it as is, unless you feel strongly about this, in which case I can look to add it.

Reply:
Hey @Charl1996 

> the modules is filtered out

That was likely because the app doc was wrapped with `SavedAppBuild` which you can ignore for your use case. You can use the raw doc or `wrap_app` function to wrap it correctly to an application doc that you can then use or to fetch the app builds using the ID or the version number.

Let me know if you want to pair on this.

Reply:
I did swap out the wrapper class with `wrap_app`, but the issue does not reside with the wrapper - it's with the couch view.

Reply:
Okay, thanks for checking that out, looks like we modify what we save.

Though it should have the ID of the build which you can then use to get the app using `get_app`.
Using this couch view, if it works for this use case, would just prevent you from loading all build ids and then handling the indexing to get the previous version.
In terms of performance I am not certain if this is better than what you already have but logically it seemed better to have db just sort things out and give back exactly what we need instead of us using indexing to find it.

Reply:
That's a good point. With your suggested approach it should clean up the code as well. I'll play around and see what it gets to.

Reply:
See https://github.com/dimagi/commcare-hq/pull/35742